app.js

```diff
@@ -365,3 +368,94 @@ async function saveFileToDisk(res, address) {
     writeStream.on('error', reject);
   });
 }
+
+/**
+ * Guesses the real extension of a file if needed
+ *
+ * @param downloadResult Result of the download
+ */
+async function ensureFileTypeIsCorrect(downloadResult) {
+  const contentType = downloadResult.result.headers.get('content-type');
+
+  // If content type in binary or if we didn't find an extension yet, try guessing
+  if (contentType == 'application/octet-stream' || !downloadResult.extension) {
+    const guessedExtension = await guessRealExtension(downloadResult.cachedFileAddress);
+
+    if (guessedExtension && (guessedExtension != downloadResult.extension)) {
+      const updatedResult = await updateFileExtension(downloadResult.cachedFileAddress, guessedExtension);
+      downloadResult.cachedFileAddress = updatedResult.cachedFileAddress;
+      downloadResult.cachedFileName = updatedResult.cachedFileName;
+      downloadResult.extension = guessedExtension;
+    }
+  }
+
+  return downloadResult;
+}
+
+/**
+ * Try deducing file extension using magic numbers and parsing
+ *
+ * @param fileAddress Location of the saved file
+ */
+async function guessRealExtension(fileAddress) {
+  const fileType = await FileType.fromFile(fileAddress)
+  if (fileType) {
+    // File type can be deduced from magic numbers
+    return `.${fileType.ext}`;
+  } else {
+    const bufferedFile = fs.readFileSync(fileAddress, 'utf8');
+    const isTextFile = isText(null, bufferedFile);
+
+    if (isTextFile) {
+      // File is a text file, guessing if it's html, else default text format
+      const doctype = getHtmlDoctypeFromBuffer(bufferedFile);
+      if (doctype) {
+        return '.html';
+      } else {
+        return DEFAULT_TEXT_FORMAT;
+      }
+    }
+  }
+
+  // Default to .bin
+  return '.bin';
+}
+
+/**
+ * Checks if a document has an html doctype
+ *
+ * @param {Buffer} bufferedFile The buffered file to parse
+ */
+function getHtmlDoctypeFromBuffer(bufferedFile) {
+  // In HTML, the doctype is mandatory. Other tags such as <html>, <head>, ... can
+  // be omitted in certain circomstances, making it hard to use to determine if a file's
+  // content is HTML. See also https://html.spec.whatwg.org/dev/syntax.html#syntax
+  const htmlparser2 = require("htmlparser2");
+  const document = htmlparser2.parseDOM(bufferedFile);
+
+  if (document) {
+    return document.find(element => {
+      const isDoctype = element.name ? element.name.toLowerCase() == '!doctype' : false;
+      const isHtml = element.nodeValue ? element.nodeValue.includes('html') : false;
```

Comment:
I think it is too strict. I happens often we don't have fully spec compliant html we get. But browsers (i.e user downloads the cached file and opens it in it's browser) are very tolerant. So geuss, if we detect a tag or something (onCloseTag) we're good to go. 
Yes it will be overly enthusiastic, label probs other stuff as html too. We'll need to refine later.

Reply:
Also, wrap it in try catch, en return false if you caught an error for now.